src/ir/local-graph.h

```diff
@@ -51,18 +57,41 @@ struct LocalGraph {
   // for a param.
   //
   // Often there is a single set, or a phi or two items, so we use a small set.
+  //
+  // If this LocalGraph is in lazy mode then this may perform the computation
+  // of sets at this time. As a result, the returned Sets from this function
+  // should be considered invalidated by other calls to getSets (the same as if
+  // you read from a std::unordered_map and then modified that map). The const
+  // version of this method has no such issue, of course.
   using Sets = SmallSet<LocalSet*, 2>;
-  const Sets& getSets(LocalGet* get) const {
-    // When we return an empty result, use a canonical constant empty set to
-    // avoid allocation.
-    static const Sets empty;
+  const Sets& getSets(LocalGet* get) {
     auto iter = getSetsMap.find(get);
     if (iter == getSetsMap.end()) {
-      return empty;
+      if (mode == Mode::Lazy) {
+        // In lazy mode, a missing entry means we did not do the computation
+        // yet. Do it now.
+        computeGetSets(get);
+        iter = getSetsMap.find(get);
+        assert(iter != getSetsMap.end());
+      } else {
+        // In eager mode, a missing entry means there is nothing there (and we
+        // saved a little space by not putting something there).
+        //
+        // Use a canonical constant empty set to avoid allocation.
+        static const Sets empty;
+        return empty;
+      }
     }
     return iter->second;
   }
 
+  const Sets& getSets(LocalGet* get) const {
+    // In eager mode, the non-const version makes no changes, so we can call it
+    // from this const one.
+    assert(mode == Mode::Eager);
+    return const_cast<LocalGraph*>(this)->getSets(get);
```

Comment:
Might as well make the `getSetsMap` `mutable` so it can be modified in lazy mode via a constant reference. That will make the public API of lazy and eager modes indistinguishable and will remove the foot gun of accidentally having an unusable constant reference to a lazy `LocalGraph`.

Reply:
Good idea, thanks, done.